Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2024 Fangming Li. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fangming Li, Jujian Zhang
+-/
+import Mathlib.Algebra.Polynomial.AlgebraMap
+import Mathlib.Algebra.Polynomial.Eval.SMul
+import Mathlib.RingTheory.Polynomial.Pochhammer
+import Mathlib.RingTheory.PowerSeries.WellKnown
+import Mathlib.Tactic.FieldSimp
+
+/-!
+# Hilbert polynomials
+
+In this file, we formalise the following statement: if `F` is a field with characteristic `0`, then
+given any `p : F[X]` and `d : ℕ`, there exists some `h : F[X]` such that for any large enough
+`n : ℕ`, `h(n)` is equal to the coefficient of `Xⁿ` in the power series expansion of `p/(1 - X)ᵈ`.
+This `h` is unique and is denoted as `Polynomial.hilbert p d`.
+
+## Main definitions
+
+* `Polynomial.hilbert p d`. Given a field `F`, a polynomial `p : F[X]` and a natural number `d`, if
+  `F` is of characteristic `0`, then `Polynomial.hilbert p d : F[X]` is the polynomial whose value
+  at `n` equals the coefficient of `Xⁿ` in the power series expansion of `p/(1 - X)ᵈ`.
+
+## TODO
+
+* Hilbert polynomials of finitely generated graded modules over Noetherian rings.
+-/
+
+open Nat PowerSeries
+
+variable (F : Type*) [Field F]
+
+namespace Polynomial
+
+/--
+For any field `F` and natrual numbers `d` and `k`, `Polynomial.preHilbert F d k` is defined as
+`(d.factorial : F)⁻¹ • ((ascPochhammer F d).comp (X - (C (k : F)) + 1))`. This is the most basic
+form of Hilbert polynomials. `Polynomial.preHilbert ℚ d 0` is exactly the Hilbert polynomial of
+the polynomial ring `ℚ[X_0,...,X_d]` viewed as a graded module over itself. See also the theorem
```

Comment:
Is it also true that `Polynomial.preHilbert F d k` is equal to `Polynomial.hilbert` of something? Perhaps you should mention that too.